File: src/main/java/com/joecollins/graphics/components/SwingometerFrameBuilder.java
package com.joecollins.graphics.components;

import com.joecollins.bindings.Bindable;
import com.joecollins.bindings.BindableList;
import com.joecollins.bindings.Binding;
import com.joecollins.bindings.BindingReceiver;
import com.joecollins.bindings.IndexedBinding;
import java.awt.Color;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;
import org.apache.commons.lang3.tuple.Pair;

public class SwingometerFrameBuilder {

  private static class Properties extends Bindable<Properties.Property> {

    enum Property {
      VALUE
    }

    private Number value = 0;
    private Number max = 1;
    private Number bucketSize = 1;
    private Number tickInterval = null;

    public void setValue(Number value) {
      this.value = Double.isNaN(value.doubleValue()) ? 0.0 : value;
      onPropertyRefreshed(Properties.Property.VALUE);
    }

    public void setMax(Number max) {
      this.max = max;
      onPropertyRefreshed(Properties.Property.VALUE);
    }

    public void setBucketSize(Number bucketSize) {
      this.bucketSize = bucketSize;
      onPropertyRefreshed(Properties.Property.VALUE);
    }

    public void setTickInterval(Number tickInterval) {
      this.tickInterval = tickInterval;
      onPropertyRefreshed(Properties.Property.VALUE);
    }
  }

  private SwingometerFrame frame = new SwingometerFrame();
  private final Properties properties = new Properties();

  public static SwingometerFrameBuilder basic(
      Binding<? extends Pair<Color, Color>> colors, Binding<? extends Number> value) {
    BindingReceiver<? extends Pair<Color, Color>> colorsRec = new BindingReceiver<>(colors);
    SwingometerFrameBuilder builder = new SwingometerFrameBuilder();
    value.bindLegacy(builder.properties::setValue);
    builder.frame.setLeftColorBinding(colorsRec.getBinding(Pair::getLeft));
    builder.frame.setRightColorBinding(colorsRec.getBinding(Pair::getRight));
    builder.frame.setValueBinding(
        Binding.propertyBinding(
            builder.properties, props -> props.value, Properties.Property.VALUE));
    return builder;
  }

  public SwingometerFrameBuilder withRange(Binding<? extends Number> range) {
    range.bindLegacy(properties::setMax);
    frame.setRangeBinding(
        Binding.propertyBinding(properties, this::getMax, Properties.Property.VALUE));
    return this;
  }

  private double getMax(Properties props) {
    return Math.max(
        props.max.doubleValue(),
        props.bucketSize.doubleValue()
            * Math.ceil(Math.abs(props.value.doubleValue() / props.bucketSize.doubleValue())));
  }

  public SwingometerFrameBuilder withBucketSize(Binding<? extends Number> bucketSize) {
    bucketSize.bindLegacy(properties::setBucketSize);
    frame.setNumBucketsPerSideBinding(
        Binding.propertyBinding(
            properties,
            props -> (int) Math.round(getMax(props) / props.bucketSize.doubleValue()),
            Properties.Property.VALUE));
    return this;
  }

  private class Tick {
    private final double level;
    private final String text;

    public Tick(double level, String text) {
      this.level = level;
      this.text = text;
    }
  }

  public SwingometerFrameBuilder withTickInterval(
      Binding<? extends Number> tickInterval, Function<? super Number, String> tickStringFunc) {
    BindableList<Tick> ticks = new BindableList<>();
    tickInterval.bindLegacy(properties::setTickInterval);
    Binding.propertyBinding(
            properties, props -> getTicks(props, tickStringFunc), Properties.Property.VALUE)
        .bindLegacy(ticks::setAll);
    frame.setNumTicksBinding(Binding.sizeBinding(ticks));
    frame.setTickPositionBinding(IndexedBinding.propertyBinding(ticks, t -> t.level));
    frame.setTickTextBinding(IndexedBinding.propertyBinding(ticks, t -> t.text));
    return this;
  }

  private List<Tick> getTicks(Properties props, Function<? super Number, String> tickStringFunc) {
    ArrayList<Tick> ticks = new ArrayList<>();
    double max = getMax(props);
    ticks.add(new Tick(0, tickStringFunc.apply(0)));
    for (double i = props.tickInterval.doubleValue();
        i < max;
        i += props.tickInterval.doubleValue()) {
      ticks.add(new Tick(i, tickStringFunc.apply(i)));
      ticks.add(new Tick(-i, tickStringFunc.apply(i)));
    }
    return ticks;
  }

  public SwingometerFrameBuilder withLeftNeedingToWin(Binding<? extends Number> leftToWin) {
    frame.setLeftToWinBinding(leftToWin);
    return this;
  }

  public SwingometerFrameBuilder withRightNeedingToWin(Binding<? extends Number> rightToWin) {
    frame.setRightToWinBinding(rightToWin);
    return this;
  }

  public <T> SwingometerFrameBuilder withOuterLabels(
      BindableList<T> labels,
      Function<T, ? extends Number> positionFunc,
      Function<T, String> labelFunc,
      Function<T, Color> colorFunc) {
    frame.setNumOuterLabelsBinding(Binding.sizeBinding(labels));
    frame.setOuterLabelPositionBinding(IndexedBinding.propertyBinding(labels, positionFunc::apply));
    frame.setOuterLabelTextBinding(IndexedBinding.propertyBinding(labels, labelFunc::apply));
    frame.setOuterLabelColorBinding(IndexedBinding.propertyBinding(labels, colorFunc::apply));
    return this;
  }

  public <T> SwingometerFrameBuilder withDots(
      BindableList<T> dots,
      Function<T, ? extends Number> positionFunc,
      Function<T, Color> colorFunc) {
    return withDots(dots, positionFunc, colorFunc, d -> "");
  }

  public <T> SwingometerFrameBuilder withDots(
      BindableList<T> dots,
      Function<T, ? extends Number> positionFunc,
      Function<T, Color> colorFunc,
      Function<T, String> labelFunc) {
    return withDots(dots, positionFunc, colorFunc, labelFunc, d -> true);
  }

  public <T> SwingometerFrameBuilder withDotsSolid(
      BindableList<T> dots,
      Function<T, ? extends Number> positionFunc,
      Function<T, Color> colorFunc,
      Predicate<T> solidFunc) {
    return withDots(dots, positionFunc, colorFunc, d -> "", solidFunc);
  }

  public <T> SwingometerFrameBuilder withDots(
      BindableList<T> dots,
      Function<T, ? extends Number> positionFunc,
      Function<T, Color> colorFunc,
      Function<T, String> labelFunc,
      Predicate<T> solidFunc) {
    frame.setNumDotsBinding(Binding.sizeBinding(dots));
    frame.setDotsPositionBinding(IndexedBinding.propertyBinding(dots, positionFunc::apply));
    frame.setDotsColorBinding(IndexedBinding.propertyBinding(dots, colorFunc::apply));
    frame.setDotsLabelBinding(IndexedBinding.propertyBinding(dots, labelFunc::apply));
    frame.setDotsSolidBinding(IndexedBinding.propertyBinding(dots, solidFunc::test));
    return this;
  }

  public <T> SwingometerFrameBuilder withFixedDots(
      List<T> dots,
      Function<T, ? extends Number> positionFunc,
      Function<T, Binding<Color>> colorFunc) {
    return withFixedDots(dots, positionFunc, colorFunc, d -> "");
  }

  public <T> SwingometerFrameBuilder withFixedDots(
      List<T> dots,
      Function<T, ? extends Number> positionFunc,
      Function<T, Binding<Color>> colorFunc,
      Function<T, String> labelFunc) {
    return withFixedDots(dots, positionFunc, colorFunc, labelFunc, d -> true);
  }

  public <T> SwingometerFrameBuilder withFixedDotsSolid(
      List<T> dots,
      Function<T, ? extends Number> positionFunc,
      Function<T, Binding<Color>> colorFunc,
      Predicate<T> solidFunc) {
    return withFixedDots(dots, positionFunc, colorFunc, d -> "", solidFunc);
  }

  public <T> SwingometerFrameBuilder withFixedDots(
      List<T> dots,
      Function<T, ? extends Number> positionFunc,
      Function<T, Binding<Color>> colorFunc,
      Function<T, String> labelFunc,
      Predicate<T> solidFunc) {
    frame.setNumDotsBinding(Binding.fixedBinding(dots.size()));
    frame.setDotsPositionBinding(
        IndexedBinding.listBinding(dots, d -> () -> positionFunc.apply(d)));
    frame.setDotsColorBinding(IndexedBinding.listBinding(dots, d -> colorFunc.apply(d)));
    frame.setDotsLabelBinding(IndexedBinding.listBinding(dots, d -> () -> labelFunc.apply(d)));
    frame.setDotsSolidBinding(IndexedBinding.listBinding(dots, d -> () -> solidFunc.test(d)));
    return this;
  }

  public SwingometerFrameBuilder withHeader(Binding<String> header) {
    frame.setHeaderBinding(header);
    return this;
  }

  public SwingometerFrame build() {
    return frame;
  }
}

File: src/main/java/com/joecollins/graphics/components/SwingometerFrameBuilder.kt
package com.joecollins.graphics.components

import com.joecollins.bindings.Bindable
import com.joecollins.bindings.Binding
import com.joecollins.bindings.BindingReceiver
import com.joecollins.bindings.mapElements
import com.joecollins.pubsub.asOneTimePublisher
import java.awt.Color
import java.util.ArrayList
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.roundToInt

class SwingometerFrameBuilder {
    private class Properties : Bindable<Properties, Properties.Property>() {
        enum class Property {
            VALUE
        }

        private var _value: Number = 0
        private var _max: Number = 1
        private var _bucketSize: Number = 1
        private var _tickInterval: Number = 1

        var value: Number
            get() { return _value }
            set(value) {
                _value = if (java.lang.Double.isNaN(value.toDouble())) 0.0 else value
                onPropertyRefreshed(Property.VALUE)
            }

        var max: Number
            get() { return _max }
            set(max) {
                this._max = max
                onPropertyRefreshed(Property.VALUE)
            }

        var bucketSize: Number
            get() { return _bucketSize }
            set(bucketSize) {
                this._bucketSize = bucketSize
                onPropertyRefreshed(Property.VALUE)
            }

        var tickInterval: Number
            get() { return _tickInterval }
            set(tickInterval) {
                this._tickInterval = tickInterval
                onPropertyRefreshed(Property.VALUE)
            }
    }

    private var rangeBinding: Binding<Double>? = null
    private var numBucketsPerSideBinding: Binding<Int>? = null
    private var ticksBinding: Binding<List<SwingometerFrame.Tick>>? = null
    private var leftToWinBinding: Binding<Number>? = null
    private var rightToWinBinding: Binding<Number>? = null
    private var outerLabelsBinding: Binding<List<SwingometerFrame.OuterLabel>>? = null
    private var dotsBinding: Binding<List<SwingometerFrame.Dot>>? = null
    private var headerBinding: Binding<String?>? = null
    private var leftColorBinding: Binding<Color>? = null
    private var rightColorBinding: Binding<Color>? = null
    private var valueBinding: Binding<Number>? = null

    private val properties = Properties()

    fun withRange(range: Binding<Number>): SwingometerFrameBuilder {
        range.bind { max -> properties.max = max }
        this.rangeBinding = Binding.propertyBinding(properties, { getMax(it) }, Properties.Property.VALUE)
        return this
    }

    private fun getMax(props: Properties): Double {
        return max(
            props.max.toDouble(),
            props.bucketSize.toDouble() *
                ceil(abs(props.value.toDouble() / props.bucketSize.toDouble()))
        )
    }

    fun withBucketSize(bucketSize: Binding<Number>): SwingometerFrameBuilder {
        bucketSize.bind { properties.bucketSize = it }
        this.numBucketsPerSideBinding = Binding.propertyBinding(
            properties,
            { (getMax(it) / it.bucketSize.toDouble()).roundToInt() },
            Properties.Property.VALUE
        )
        return this
    }

    private inner class Tick(val level: Double, val text: String)

    fun withTickInterval(
        tickInterval: Binding<Number>,
        tickStringFunc: (Number) -> String
    ): SwingometerFrameBuilder {
        tickInterval.bind { properties.tickInterval = it }
        val ticks = Binding.propertyBinding(properties, { getTicks(it, tickStringFunc) }, Properties.Property.VALUE)
        this.ticksBinding = ticks.mapElements {
            SwingometerFrame.Tick(it.level, it.text)
        }
        return this
    }

    private fun getTicks(props: Properties, tickStringFunc: (Number) -> String): List<Tick> {
        val ticks = ArrayList<Tick>()
        val max = getMax(props)
        ticks.add(Tick(0.0, tickStringFunc(0)))
        var i = props.tickInterval.toDouble()
        while (i < max) {
            ticks.add(Tick(i, tickStringFunc(i)))
            ticks.add(Tick(-i, tickStringFunc(i)))
            i += props.tickInterval.toDouble()
        }
        return ticks
    }

    fun withLeftNeedingToWin(leftToWin: Binding<Number>): SwingometerFrameBuilder {
        this.leftToWinBinding = leftToWin
        return this
    }

    fun withRightNeedingToWin(rightToWin: Binding<Number>): SwingometerFrameBuilder {
        this.rightToWinBinding = rightToWin
        return this
    }

    fun <T> withOuterLabels(
        labels: Binding<List<T>>,
        positionFunc: (T) -> Number,
        labelFunc: (T) -> String,
        colorFunc: (T) -> Color
    ): SwingometerFrameBuilder {
        this.outerLabelsBinding = labels.mapElements {
            SwingometerFrame.OuterLabel(positionFunc(it), labelFunc(it), colorFunc(it))
        }
        return this
    }

    fun <T> withDots(
        dots: Binding<List<T>>,
        positionFunc: (T) -> Number,
        colorFunc: (T) -> Color
    ): SwingometerFrameBuilder {
        return withDots(dots, positionFunc, colorFunc, { "" })
    }

    fun <T> withDots(
        dots: Binding<List<T>>,
        positionFunc: (T) -> Number,
        colorFunc: (T) -> Color,
        labelFunc: (T) -> String
    ): SwingometerFrameBuilder {
        return withDots(dots, positionFunc, colorFunc, labelFunc, { true })
    }

    fun <T> withDotsSolid(
        dots: Binding<List<T>>,
        positionFunc: (T) -> Number,
        colorFunc: (T) -> Color,
        solidFunc: (T) -> Boolean
    ): SwingometerFrameBuilder {
        return withDots(dots, positionFunc, colorFunc, { "" }, solidFunc)
    }

    fun <T> withDots(
        dots: Binding<List<T>>,
        positionFunc: (T) -> Number,
        colorFunc: (T) -> Color,
        labelFunc: (T) -> String,
        solidFunc: (T) -> Boolean
    ): SwingometerFrameBuilder {
        this.dotsBinding =
            dots.mapElements { SwingometerFrame.Dot(positionFunc(it), colorFunc(it), labelFunc(it), solidFunc(it)) }
        return this
    }

    fun <T> withFixedDots(
        dots: List<T>,
        positionFunc: (T) -> Number,
        colorFunc: (T) -> Binding<Color>
    ): SwingometerFrameBuilder {
        return withFixedDots(dots, positionFunc, colorFunc, { "" })
    }

    fun <T> withFixedDots(
        dots: List<T>,
        positionFunc: (T) -> Number,
        colorFunc: (T) -> Binding<Color>,
        labelFunc: (T) -> String
    ): SwingometerFrameBuilder {
        return withFixedDots(dots, positionFunc, colorFunc, labelFunc, { true })
    }

    fun <T> withFixedDotsSolid(
        dots: List<T>,
        positionFunc: (T) -> Number,
        colorFunc: (T) -> Binding<Color>,
        solidFunc: (T) -> Boolean
    ): SwingometerFrameBuilder {
        return withFixedDots(dots, positionFunc, colorFunc, { "" }, solidFunc)
    }

    fun <T> withFixedDots(
        dots: List<T>,
        positionFunc: (T) -> Number,
        colorFunc: (T) -> Binding<Color>,
        labelFunc: (T) -> String,
        solidFunc: (T) -> Boolean
    ): SwingometerFrameBuilder {
        this.dotsBinding = Binding.listBinding(
            dots.map {
                colorFunc(it).map { c -> SwingometerFrame.Dot(positionFunc(it), c, labelFunc(it), solidFunc(it)) }
            }
        )
        return this
    }

    fun withHeader(header: Binding<String?>): SwingometerFrameBuilder {
        this.headerBinding = header
        return this
    }

    fun build(): SwingometerFrame {
        return SwingometerFrame(
            headerPublisher = headerBinding?.toPublisher() ?: (null as String?).asOneTimePublisher(),
            rangePublisher = rangeBinding?.toPublisher() ?: 1.asOneTimePublisher(),
            valuePublisher = valueBinding?.toPublisher() ?: 0.asOneTimePublisher(),
            leftColorPublisher = leftColorBinding?.toPublisher() ?: Color.BLACK.asOneTimePublisher(),
            rightColorPublisher = rightColorBinding?.toPublisher() ?: Color.BLACK.asOneTimePublisher(),
            numBucketsPerSidePublisher = numBucketsPerSideBinding?.toPublisher() ?: 1.asOneTimePublisher(),
            dotsPublisher = dotsBinding?.toPublisher() ?: emptyList<SwingometerFrame.Dot>().asOneTimePublisher(),
            leftToWinPublisher = leftToWinBinding?.toPublisher(),
            rightToWinPublisher = rightToWinBinding?.toPublisher(),
            ticksPublisher = ticksBinding?.toPublisher(),
            outerLabelsPublisher = outerLabelsBinding?.toPublisher()
        )
    }

    companion object {
        @JvmStatic fun basic(
            colors: Binding<Pair<Color, Color>>,
            value: Binding<Number>
        ): SwingometerFrameBuilder {
            val colorsRec: BindingReceiver<Pair<Color, Color>> = BindingReceiver(colors)
            val builder = SwingometerFrameBuilder()
            value.bind { builder.properties.value = it }
            builder.leftColorBinding = colorsRec.getBinding { it.first }
            builder.rightColorBinding = colorsRec.getBinding { it.second }
            builder.valueBinding = Binding.propertyBinding(
                builder.properties, { props: Properties -> props.value }, Properties.Property.VALUE
            )
            return builder
        }
    }
}
